Source/Tests/LinuxSyscalls/SyscallsSMCTracking.cpp

```diff
@@ -0,0 +1,326 @@
+    rv.Entry = Entry->second.Resource ? Entry->second.Resource->AOTIRCacheEntry : nullptr;
+    rv.Offset = Entry->second.Base - Entry->second.Offset;
+  }
+
+  return rv;
+}
+
+// MMan Tracking
+static std::string get_fdpath(int fd) {
+  std::error_code ec;
+  return std::filesystem::canonical(std::filesystem::path("/proc/self/fd") / std::to_string(fd), ec).string();
+}
+
+void SyscallHandler::TrackMmap(uintptr_t Base, uintptr_t Size, int Prot, int Flags, int fd, off_t Offset) {
+	Size = FEXCore::AlignUp(Size, FHU::FEX_PAGE_SIZE);
+
+  {
+    FHU::ScopedSignalMaskWithUniqueLock lk(_SyscallHandler->VMATracking.Mutex);
+
+    static uint64_t AnonSharedId = 1;
+
+    MappedResource *Resource = nullptr;
+
+    if (!(Flags & MAP_ANONYMOUS)) {
+      struct stat64 buf;
+      fstat64(fd, &buf);
+      MRID mrid{buf.st_dev, buf.st_ino};
+
+      auto ResourceInserted = VMATracking.MappedResources.insert({mrid, {nullptr, nullptr, 0}});
+      Resource = &ResourceInserted.first->second;
+
+      if (ResourceInserted.second) {
+        auto filename = get_fdpath(fd);
+        Resource->AOTIRCacheEntry = FEXCore::Context::LoadAOTIRCacheEntry(CTX, filename);
+        Resource->Iterator = ResourceInserted.first;
+      }
+
+    } else if (Flags & MAP_SHARED) {
+      MRID mrid{MRID_ANON, AnonSharedId++};
+
+      auto ResourceInserted = VMATracking.MappedResources.insert({mrid, {nullptr, nullptr, 0}});
+      LOGMAN_THROW_A_FMT(ResourceInserted.second, "VMA tracking error");
+      Resource = &ResourceInserted.first->second;
+      Resource->Iterator = ResourceInserted.first;
+    } else {
+      Resource = nullptr;
+    }
+
+    VMATracking.SetUnsafe(CTX, Resource, Base, Offset, Size, VMAOptions{Prot & PROT_WRITE, Flags & MAP_SHARED});
+  }
+
+  if (SMCChecks != FEXCore::Config::CONFIG_SMC_NONE) {
+    FEXCore::Context::InvalidateGuestCodeRange(CTX, (uintptr_t)Base, Size);
+  }
+}
+
+void SyscallHandler::TrackMunmap(uintptr_t Base, uintptr_t Size) {
+	Size = FEXCore::AlignUp(Size, FHU::FEX_PAGE_SIZE);
+
+  {
+    FHU::ScopedSignalMaskWithUniqueLock lk(_SyscallHandler->VMATracking.Mutex);
+
+    VMATracking.ClearUnsafe(CTX, Base, Size);
+  }
+
+  if (SMCChecks != FEXCore::Config::CONFIG_SMC_NONE) {
+    FEXCore::Context::InvalidateGuestCodeRange(CTX, (uintptr_t)Base, Size);
+  }
+}
+
+void SyscallHandler::TrackMprotect(uintptr_t Base, uintptr_t Size, int Prot) {
+	Size = FEXCore::AlignUp(Size, FHU::FEX_PAGE_SIZE);
+
+  {
+    FHU::ScopedSignalMaskWithUniqueLock lk(_SyscallHandler->VMATracking.Mutex);
+
+    VMATracking.ChangeUnsafe(Base, Size, VMAProt{(Prot & PROT_WRITE) != 0});
+  }
+
+  if (SMCChecks != FEXCore::Config::CONFIG_SMC_NONE) {
+    FEXCore::Context::InvalidateGuestCodeRange(CTX, Base, Size);
+  }
+}
+
+void SyscallHandler::TrackMremap(uintptr_t OldAddress, size_t OldSize, size_t NewSize, int flags, uintptr_t NewAddress) {
+	OldSize = FEXCore::AlignUp(OldSize, FHU::FEX_PAGE_SIZE);
+	NewSize = FEXCore::AlignUp(NewSize, FHU::FEX_PAGE_SIZE);
```

Comment:
Looks okay here, maybe you have tab size set wrongly?

Reply:
We use spaces not tabs, and this is tabs.

Reply:
The codebase is sprinkled all over with tabs, is this something we've actually cared about so far?

I have my editor setup to convert tabs to spaces, so this is from some copy paste that didn't convert, most likely.

Reply:
And if this is something we actually want to enforce, let's setup a CI for it. Otherwise things will just slip through.

Reply:
This is something we care about, each time I encounter this I force reformat the file.
I agree that we should have CI enforcing this.

Reply:
Considering that 100s of my PRs have been merged, and for several months I actually hadn't the tab fixed editor so many of my commits had a tab size of 4, and that this has never been brought up before, I don't know, :shrug:

Reply:
I've let it slide the entire time because I'm not pedantic enough to comment about whitespace being incorrect.
Since it was commented on that you didn't see what was wrong, I brought it up.

Reply:
Figured that, though let's setup some strict guidelines on these things if we do care about them. A rule regularly non enforced is a non existent rule.

Reply:
Agreed on the CI bit (and another reason to push clang-format again). The parts of the code base I've seen so far consistently used spaces for indentation, so I just commented because tabs stand out in GitHub's code diff and I thought it was an isolated instance :)